doc/source/serve/tutorials/gradio-integration.md

```diff
@@ -0,0 +1,122 @@
+```
+
+
+### Understanding `GradioServer`
+In order to deploy your Gradio app onto Ray Serve, you need to wrap your Gradio app in a Serve [deployment](serve-key-concepts-deployment). `GradioServer` acts as that wrapper. It serves your Gradio app on a Ray cluster.
+> **NOTE** `GradioServer` is simply `GradioIngress` but wrapped in a Serve deployment.
+```{literalinclude} ../../../../python/ray/serve/gradio.py
+:start-after: __doc_gradio_ingress_begin__
+:end-before: __doc_gradio_ingress_end__
+```
+
+### Deploy your Gradio Server
+Replicas in a deployment are copies of your program living on the Ray cluster, and more replicas means your deployment can serve more client requests. You can increase the number of replicas of your application or increase the number of CPUs and/or GPUs available to each replica.
+
+Then, using either the example we created above, or an existing Gradio app (of type `Interface`, `Block`, `Parallel`, etc.), wrap it in your Gradio Server.
+
+```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-integration.py
+:start-after: __doc_app_begin__
+:end-before: __doc_app_end__
+```
+
+Finally, deploy your Gradio Server! Run the following in your terminal:
+```console
+$ serve run demo:app
+```
+
+Now you can access your Gradio app at `http://localhost:8000`! This is what it should look like:
+![Gradio Result](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/gradio_result.png)
+
+See [Putting Ray Serve Deployment Graphs in Production](https://docs.ray.io/en/master/serve/production.html#id1) for more information on how to deploy your app in production.
+
+
+## Complex Example: Parallelism
+You can run multiple models in parallel with Ray Serve.
+
+### Original Approach
+Suppose you want to run the following program.
+
+1. Take two text generation models, [`gpt2`](https://huggingface.co/gpt2) and [`EleutherAI/gpt-neo-125M`](https://huggingface.co/EleutherAI/gpt-neo-125M).
+2. Run the two models on the same input text, such that the generated text has a minimum length of 20 and maximum length of 100.
+3. Display the outputs of both models using Gradio.
+
+This is how you would do it normally:
+
+```{literalinclude} ../../../../python/ray/serve/examples/doc/gradio-original.py
+:start-after: __doc_code_begin__
+:end-before: __doc_code_end__
+```
+
+### Parallelize using Ray Serve
+
+With Ray Serve, we can parallelize the two text generation models by wrapping each model in a separate Ray Serve [deployment](serve-key-concepts-deployment). Deployments are defined by decorating a Python class or function with `serve.deployment`, and usually wrap the models that you want to deploy on Ray Serve and handle incoming requests.
+
+First, let's import our dependencies. Note that we need to import `GradioIngress` instead of `GradioServer` like before, since we are actually building our own deployable `MyGradioServer` with more customization that allows us to run models in parallel.
```

Comment:
```suggestion
First, let's import our dependencies. Note that we need to import `GradioIngress` instead of `GradioServer` like before since we're now building a customized `MyGradioServer` that can run models in parallel.
```